fe/fe-core/src/main/java/com/starrocks/catalog/TabletInvertedIndex.java

```diff
@@ -210,110 +142,89 @@ public void markTabletForceDelete(Tablet tablet) {
+        tabletMetaMap.remove(tabletId);
+
+        LOG.debug("delete tablet: {}", tabletId);
     }
 
     // Only for test
-    public Map<Long, Map<Long, Replica>> getReplicaMetaTable() {
-        return replicaMetaTable;
+    public ConcurrentLong2ObjectHashMap<CopyOnWriteArrayList<Replica>> getReplicaMetaTable() {
+        return tabletToReplicaList;
     }
 
     public void addReplica(long tabletId, Replica replica) {
         if (GlobalStateMgr.isCheckpointThread()) {
             return;
         }
-        writeLock();
-        try {
-            Preconditions.checkState(tabletMetaMap.containsKey(tabletId));
-            setReplica(replicaMetaTable, tabletId, replica.getBackendId(), replica);
-            replicaToTabletMap.put(replica.getId(), tabletId);
-            setReplica(backingReplicaMetaTable, replica.getBackendId(), tabletId, replica);
-            LOG.debug("add replica {} of tablet {} in backend {}",
-                    replica.getId(), tabletId, replica.getBackendId());
-        } finally {
-            writeUnlock();
-        }
+
+        CopyOnWriteArrayList<Replica> replicas = tabletToReplicaList.computeIfAbsent(tabletId, k -> new CopyOnWriteArrayList<>());
+        replicas.add(replica);
+
+        Set<Long> tabletIds = backendToTabletIdList.computeIfAbsent(replica.getBackendId(), k -> ConcurrentHashMap.newKeySet());
+        tabletIds.add(tabletId);
+
+        LOG.debug("add replica {} of tablet {} in backend {}", replica.getId(), tabletId, replica.getBackendId());
     }
 
     public void deleteReplica(long tabletId, long backendId) {
         if (GlobalStateMgr.isCheckpointThread()) {
             return;
         }
-        writeLock();
-        try {
-            if (!tabletMetaMap.containsKey(tabletId)) {
-                return;
-            }
-            if (replicaMetaTable.containsKey(tabletId)) {
-                Replica replica = removeReplica(replicaMetaTable, tabletId, backendId);
-                Preconditions.checkState(replica != null);
-                replicaToTabletMap.remove(replica.getId());
-                removeReplica(backingReplicaMetaTable, backendId, tabletId);
-                LOG.debug("delete replica {} of tablet {} in backend {}",
-                        replica.getId(), tabletId, backendId);
-            } else {
-                // this may happen when fe restart after tablet is empty(bug cause)
-                // add log instead of assertion to observe
-                LOG.error("tablet[{}] contains no replica in inverted index", tabletId);
-            }
-        } finally {
-            writeUnlock();
+        if (tabletMetaMap.get(tabletId) == null) {
+            return;
+        }
+        CopyOnWriteArrayList<Replica> replicaList = tabletToReplicaList.get(tabletId);
+        if (replicaList != null) {
+            replicaList.removeIf(replica -> replica.getBackendId() == backendId);
+        }
+
+        Set<Long> tabletIds = backendToTabletIdList.get(backendId);
```

Comment:
Modification of `tabletToReplicaList` and `backendToTabletIdList` are in two statements. Considering concurrent `deleteReplica()` and `addReplica()` for the same tablet and backendId, the final result may be inconsistent because of the two ops are not atomic.